src/js/hooks/inboundV2/addItems/useInboundAddItemsForm.jsx

```diff
@@ -261,56 +300,68 @@ const useInboundAddItemsForm = ({ next, previous }) => {
     try {
       spinner.show();
       await apiClient.post(STOCK_MOVEMENT_UPDATE_INVENTORY_ITEMS(queryParams.id), payload);
-      transitionToNextStepAndChangePage(formValues);
+      transitionToNextStep(formValues);
     } finally {
       spinner.hide();
     }
   };
 
+  const handleExpirationDateMismatch = (updatedValues, lineItems, hasNonZeroQuantity) => {
+    if (hasNonZeroQuantity) {
+      confirmAction(
+        () => updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems),
+        'react.stockMovement.confirmExpiryDateUpdate.message',
+        'This will update the expiry date across all depots in the system. Are you sure you want to proceed?',
+      );
+    } else {
+      updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems);
+    }
+  };
+
+  const handleTransition = (updatedValues, lineItems) => {
+    const updatedLineItems = updatedValues.lineItems;
+
+    const hasExpirationDateMismatch = updatedLineItems?.some(
+      (item) => item.inventoryItem && item.expirationDate !== item.inventoryItem.expirationDate,
+    );
+
+    const hasNonZeroQuantity = updatedLineItems?.some(
+      (item) => item.inventoryItem?.quantity && item.inventoryItem.quantity !== '0',
+    );
+
+    if (hasExpirationDateMismatch) {
+      handleExpirationDateMismatch(updatedValues, lineItems, hasNonZeroQuantity);
+    } else {
+      transitionToNextStep(updatedValues);
+    }
+  };
+
   const saveAndTransitionToNextStep = async (values, lineItems) => {
     try {
       spinner.show();
-
       const resp = await saveRequisitionItemsInCurrentStep(lineItems);
-      const updatedValues = resp ? {
-        ...values,
-        lineItems: resp.data.data.lineItems,
-      } : values;
-      const updatedLineItems = updatedValues.lineItems;
-
-      const hasExpirationDateMismatch = updatedLineItems?.some((item) =>
-        item.inventoryItem && item.expirationDate !== item.inventoryItem.expirationDate);
-
-      const hasNonZeroQuantity = updatedLineItems?.some((item) =>
-        item.inventoryItem?.quantity && item.inventoryItem.quantity !== '0');
-
-      if (hasExpirationDateMismatch) {
-        if (hasNonZeroQuantity) {
-          confirmInventoryItemExpirationDateUpdate(() =>
-            updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems));
-        } else {
-          updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems);
-        }
-      } else {
-        transitionToNextStepAndChangePage(updatedValues);
-      }
+      const updatedValues = resp
+        ? { ...values, lineItems: resp.data.data.lineItems }
+        : values;
+
+      handleTransition(updatedValues, lineItems);
     } finally {
       spinner.hide();
     }
   };
 
+  const getItemsMapByProductCode = (lineItems) => lineItems.reduce((acc, item) => {
+    const { productCode } = item.product;
+    acc[productCode] = acc[productCode] || [];
+    acc[productCode].push(item);
```

Comment:
this is not how we should use `.reduce` function.
 It's immutable function, and why you did in two lines above is actually mutating the state.
Remind yourself the interview :smile: you've had almost the same exercise to do
What needs to be changed is to remove the two lines above and to mutate the state inside the return.